### PR TITLE
unix, windows: be consistent when binding implictly in UDP

### PR DESCRIPTION
**Pull request by saghul**
Only these functions will trigger an implicit binding of a UDP handle:
- uv_udp_send
- uv_udp_recv_start
- uv_udp_set_membership

All other functions will return UV_EBADF in case the socket was not
bound.

Note: currently the socket is created and bound at the same time. This
may change in the future.

This is the last piece for https://github.com/joyent/libuv/issues/1205

/cc @indutny

### CODE REVIEW

**Comment by saghul** [comment]
I'll add some tests after first review.

**Comment by indutny** [comment]
Looks good

**Comment by saghul** [comment]
Updated with tests.

File: test/test-udp-options.c [hunk 1]
```diff
@@ -86,3 +86,38 @@ TEST_IMPL(udp_options) {
...
+
+TEST_IMPL(udp_no_autobind) {
+  uv_loop_t* loop;
+  uv_udp_t h;
+  int r;
+
+  loop = uv_default_loop();
+
+  r = uv_udp_init(loop, &h);
+  ASSERT(r == 0);
```
Comment (indutny) [inline test/test-udp-options.c]: Please merge with the line above.

**Comment by saghul** [comment]
Updated. Is that what you meant?

**Comment by indutny** [comment]
Not really, I meant:

``` C
ASSERT(UV_EBADF == uv_udp_set_multicast_interface(&h, "0.0.0.0"));
```

Instead of two lines.

**Comment by saghul** [comment]
Ah, I see. Updated.

File: test/test-udp-options.c [hunk 2]
```diff
@@ -86,3 +86,27 @@ TEST_IMPL(udp_options) {
...
+  ASSERT(0 == uv_udp_init(loop, &h));
+  ASSERT(UV_EBADF == uv_udp_set_multicast_ttl(&h, 32));
+  ASSERT(UV_EBADF == uv_udp_set_broadcast(&h, 1));
+  ASSERT(UV_EBADF == uv_udp_set_ttl(&h, 1));
+  ASSERT(UV_EBADF == uv_udp_set_multicast_loop(&h, 1));
+  ASSERT(UV_EBADF == uv_udp_set_multicast_interface(&h, "0.0.0.0"));
+
+  uv_close((uv_handle_t*) &h, NULL);
+
+  r = uv_run(loop, UV_RUN_DEFAULT);
```
Comment (indutny) [inline test/test-udp-options.c]: Forgot one? :)

Comment (indutny) [inline test/test-udp-options.c]: Also you don't need `int r;` declaration anymore.

**Comment by saghul** [comment]
Done.

**Comment by indutny** [comment]
LGTM, thank you!

**Comment by saghul** [comment]
Landed in https://github.com/joyent/libuv/commit/386d2141e44a0d007e6d5a8535f82f9191dc9d47, thanks Fedor!

**Pull request closed without merging**